x-pack/metricbeat/module/azure/monitor_service.go

```diff
@@ -219,6 +224,44 @@ func (service *MonitorService) GetMetricDefinitions(resourceId string, namespace
 	return metricDefinitionCollection, nil
 }
 
+func (service *MonitorService) GetMetricDefinitionsWithRetry(resource *armresources.GenericResourceExpanded, namespace string) (armmonitor.MetricDefinitionCollection, error) {
```

Comment:
Shouldn't we handle the 429 status inside the main `GetMetricDefinitions()` function?

Since `GetMetricDefinitions()` can paginate the result, handling 429s outside the function may result in reprocessing the same pages on the second call.

If the code only calls `GetMetricDefinitions` or `GetMetricDefinitionsWithRetry`, I think we probably should have just one.

Have you considered handling 429s right after the `pager.NextPage()`, do you see any downsides?

Reply:
> Since GetMetricDefinitions() can paginate the result, handling 429s outside the function may result in reprocessing the same pages on the second call.

I don't understand what you mean by this. If we have an error moving to the next page, we return:
https://github.com/elastic/beats/blob/b4b1dbcf293e8a918fb225d074c45f6b049462a2/x-pack/metricbeat/module/azure/monitor_service.go#L162-L165
So we don't go over extra pages.

> If the code only calls GetMetricDefinitions or GetMetricDefinitionsWithRetry, I think we probably should have just one. Have you considered handling 429s right after the pager.NextPage(), do you see any downsides?

I think the only downside is if we want to use `GetMetricDefinitions` again in the future without retry. Also, if we want to take care of other response errors, like 5xx errors, it seems to make sense that we take care of that in a different function.
But if you think it is better to take care of that all in the same function I can move the code.

Reply:
> I don't understand what you mean by this. If we have an error moving to the next page, we return:

- you call the `GetMetricDefinitions()`, it gets and processes successfully the first page,  get the second page and receives 429; the function returns an error.
- you wait for 5m
- you call the `GetMetricDefinitions()` again, re-processing the first page

Reply:
Ah, I see. I don't like the idea of moving all the code to one function, so maybe `getMetricDefinitions` should be refactored and return the pager. And then we add a new function like

```
func (service *MonitorService) movePager(pager, metricDefinitionCollection) {
for pager.More() {
		nextPage, err := pager.NextPage(service.context)
		if err != nil {
			return armmonitor.MetricDefinitionCollection{}, err
		}

		metricDefinitionCollection.Value = append(metricDefinitionCollection.Value, nextPage.Value...)
	}
}
```

Reply:
I cannot tell from the code if `nextPage, err := pager.NextPage(service.context)` advances to a new page, even if there is an error. Do you know @zmoog ?

Reply:
> I think the only downside is if we want to use GetMetricDefinitions again in the future without retry. Also, if we want to take care of other response errors, like 5xx errors, it seems to make sense that we take care of that in a different function.

I don't have a strong opinion on this. Since the function is called by the "with retry" version, it is not unused code, so it is fine.

I would only try to avoid the risk of reprocessing the same page again (if my concern is accurate).

Reply:
We are using this function for Storage as well:

https://github.com/elastic/beats/blob/e0675778b8eed65c6fff6a1d1911833c21563755/x-pack/metricbeat/module/azure/storage/client_helper.go#L44

Do we want to make this change in this PR as well? We would have to test storage metricset in that case as well

Reply:
I updated the code to not be going over the pages multiples times over. I tested it again and the results still yield:

![image](https://github.com/elastic/beats/assets/113898685/2ac6ce92-96d2-4cf2-bf6f-e6c5cae2db0b)